File: 18-4sum/4sum.go
func fourSum(nums []int, target int) [][]int {
    sort.Ints(nums)

    ans := make([][]int, 0)
    n := len(nums)
    for i := 0; i < n - 3; i++ {
        if i != 0 && nums[i-1] == nums[i] {
            continue
        }
        for j := i + 1; j < n - 2; j++ {
            if j != i + 1 && nums[j-1] == nums[j] {
                continue
            }
            low := j + 1
            high := n - 1
            for low < high {
                sum := nums[i] + nums[j] + nums[low] + nums[high]
                if sum < target {
                    low++
                } else if sum > target {
                    high--
                } else {
                    ans = append(ans, []int{nums[i], nums[j], nums[low], nums[high]})
                    low++
                    for low < high && nums[low-1] == nums[low] {
                        low++
                    }
                }
            }
        }
    }
    return ans
}

File: 18-4sum/4sum.cpp
class Solution {
public:
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        int n = nums.size();
        sort(nums.begin(), nums.end());
        set<vector<int>> set;
        for (int i = 0; i < n - 3; i++) {
            for(int j = i+1; j < n - 2; j++) {
                int64_t subTarget = (int64_t)target - (int64_t)nums[i] - (int64_t)nums[j];
                int low = j + 1, high = n - 1;
                while(low < high) {
                    if(nums[low] + nums[high] < subTarget) {
                        low++;
                    } else if(nums[low] + nums[high] > subTarget) {
                        high--;
                    } else {
                        set.insert({nums[i], nums[j], nums[low], nums[high]});
                        low++, high--;
                    }
                }
            }
        }

        vector<vector<int>> ans;
        for(auto item : set) {
            ans.push_back(item);
        }
        return ans;
    }
};